### PR TITLE
snap: make snaps vanishing less fatal for the system

### PR DESCRIPTION
**Pull request by mvo5**
This is a first step towards making snaps that vanish (e.g. because a `snap try` dir gets removed) less fatal to the system.

This does not do any cleanup. We need to discuss the best way to do this next. One possible way would be to make ReadInfo() return some Info with a "vanished" flag so the user can see that its in a broken state and can manually clean up. But thats just a strawman, ideas welcome.

### CODE REVIEW

File: daemon/snap.go [hunk 1]
```diff
@@ -99,6 +99,9 @@ func allLocalSnapInfos(st *state.State) ([]aboutSnap, error) {
 	for name, snapState := range snapStates {
 		info, err := snap.ReadInfo(name, snapState.CurrentSideInfo())
 		if err != nil {
+			if _, ok := err.(*snap.SnapVanishedError); ok {
```
Comment (pedronis) [inline daemon/snap.go]: there are helpers in snapstate I think that needs a similar treatment, some are used by ifacestate I think

File: snap/info.go [hunk 2]
```diff
@@ -357,12 +357,21 @@ func infoFromSnapYamlWithSideInfo(meta []byte, si *SideInfo) (*Info, error) {
 	return info, nil
 }
 
+type SnapVanishedError struct {
```
Comment (niemeyer) [inline snap/info.go]: `NotFoundError`, becoming `snap.NotFoundError`?

File: snap/info.go [hunk 3]
```diff
@@ -357,12 +357,21 @@ func infoFromSnapYamlWithSideInfo(meta []byte, si *SideInfo) (*Info, error) {
...
+	Snap     string
+	Revision Revision
+}
+
+func (e SnapVanishedError) Error() string {
+	return fmt.Sprintf("cannot find mounted snap %q at revision %s", e.Snap, e.Revision)
```
Comment (niemeyer) [inline snap/info.go]: s/mounted/installed/.. that's an implementation detail leaking into the UI.. the user has no idea about the fact we mount it.

File: daemon/snap.go (hunk 1 above)
Comment (niemeyer) [inline daemon/snap.go]: This branch looks fine as a minor step forward, but it's not clear if we should sprinkle the brittleness of this logic throughout the code.

What is the real solution, and when can we have it?  We need a place where we can find out about the broken state, and clean it up.

**Comment by niemeyer** [comment]
A couple of trivials and LGTM as a step forward, but we cannot keep it at that.

**Comment by niemeyer** [comment]
LGTM

**Pull request merged**